src/lib.rs

```diff
@@ -48,57 +48,66 @@ pub struct SwapResult {
 
 #[odra::module]
 pub struct Invariant {
-    pools: Pools,
-    tickmap: Tickmap,
-    ticks: Ticks,
-    position: Variable<Position>,
-    pool: Variable<Pool>,
-    tick: Variable<Tick>,
+    _positions: Positions,
+    _pools: Pools,
+    _tickmap: Tickmap,
+    _ticks: Ticks,
+    fee_tiers: Variable<FeeTiers>,
+    _pool_keys: Variable<PoolKeys>,
     state: Variable<State>,
-    liquidity: Variable<Liquidity>,
 }
 
 #[odra::module]
 impl Invariant {
     #[odra(init)]
-    pub fn init(&mut self) {
+    pub fn init(&mut self, protocol_fee: Percentage) {
         let caller = contract_env::caller();
-        let liquidity = Liquidity::new(U256::from(100_000_000u128));
-        let pool = Pool::default();
-        let tick = Tick::default();
-        self.liquidity.set(liquidity);
-        self.position.set(Position::default());
-        self.tick.set(tick);
-        self.pool.set(pool);
-
-        let token_0: Address = Address::Contract(ContractPackageHash::from([0x01; 32]));
-        let token_1: Address = Address::Contract(ContractPackageHash::from([0x02; 32]));
-        let fee_tier: FeeTier = FeeTier {
-            fee: Percentage::new(U128::from(1)),
-            tick_spacing: 1,
-        };
-        let pool_key: PoolKey = PoolKey::new(token_0, token_1, fee_tier).unwrap();
-        self.tickmap.flip(true, 0, 1, pool_key);
-        self.ticks.add(pool_key, 0, &tick).unwrap();
-
-        self.pools.add(pool_key, &pool).unwrap();
+
         self.state.set(State {
             admin: caller,
-            protocol_fee: Percentage::new(U128::from(10000000000u128)),
+            protocol_fee,
         })
     }
-}
+    pub fn add_fee_tier(&mut self, fee_tier: FeeTier) -> Result<(), InvariantError> {
+        let caller = contract_env::caller();
+        let state = self.state.get().unwrap();
+        let mut fee_tiers = self.fee_tiers.get_or_default();
 
-#[cfg(test)]
-mod tests {
-    use odra::test_env;
+        if caller != state.admin {
+            return Err(InvariantError::NotAdmin);
+        }
+
+        fee_tiers.add(fee_tier)?;
+
+        self.fee_tiers.set(fee_tiers);
+        Ok(())
+    }
 
-    use super::*;
-    #[test]
+    pub fn fee_tier_exist(&self, fee_tier: FeeTier) -> bool {
+        let fee_tiers = self.fee_tiers.get_or_default();
+        fee_tiers.contains(fee_tier)
+    }
+
+    pub fn remove_fee_tier(&mut self, fee_tier: FeeTier) -> Result<(), InvariantError> {
+        let caller = contract_env::caller();
+        let state = self.state.get().unwrap();
+        let mut fee_tiers = self.fee_tiers.get_or_default();
```

Comment:
It will be better to initialize empty struct inside constructor, instead of using `get_or_default` in every gets